File: configuration/src/main/java/com/puresol/commons/configuration/ConfigurationParameter.java
package com.puresol.commons.configuration;

import com.puresol.commons.math.LevelOfMeasurement;
import com.puresol.commons.math.ParameterWithArbitraryUnit;

/**
 * This class specifies a single configuration parameter.
 * 
 * @author Rick-Rainer Ludwig
 * 
 * @param <T>
 */
public class ConfigurationParameter<T> extends ParameterWithArbitraryUnit<T> {

	private static final long serialVersionUID = 969219744450680004L;

	private final String propertyKey;
	private final T defaultValue;

	public ConfigurationParameter(String name, String unit,
			LevelOfMeasurement levelOfMeasurement, String description,
			Class<T> type, String propertyKey, T defaultValue) {
		super(name, unit, levelOfMeasurement, description, type);
		this.propertyKey = propertyKey;
		this.defaultValue = defaultValue;
	}

	public final String getPropertyKey() {
		return propertyKey;
	}

	public final T getDefaultValue() {
		return defaultValue;
	}

}

File: analysis/webui/webui/src/main/webui/app/commons/configuration/ConfigurationParameter.ts
import {LevelOfMeasurement} from '../math/LevelOfMeasurement';
import {PreferencesGroup} from '../preferences/PreferencesGroup';
import {PreferencesGroupIdentifier} from '../preferences/PreferencesGroupIdentifier';

 /**
 * This class represents a single configuration parameter. The whole definition is defined here.
 */
export class ConfigurationParameter {

    /**
     * This method is used to convert a single configuration parameter into a 
     * ConfigurationParameter object.
     */
    public static fromJSON(
        preferencesGroup: PreferencesGroup,
        groupIdentifier: PreferencesGroupIdentifier,
        configurationParameter: any) {
        return new ConfigurationParameter(preferencesGroup,
            groupIdentifier,
            configurationParameter.name,
            configurationParameter.unit,
            configurationParameter.levelOfMeasurement,
            configurationParameter.description,
            configurationParameter.type,
            configurationParameter.propertyKey,
            configurationParameter.path,
            configurationParameter.defaultValue,
            String(configurationParameter.valueRepresentation));
    }

    constructor(
        public preferencesGroup: PreferencesGroup,
        public groupIdentifier: PreferencesGroupIdentifier,
        public name: string,
        public unit: string,
        public levelOfMeasurement: LevelOfMeasurement,
        public description: string,
        public type: string,
        public propertyKey: string,
        public path: string,
        public defaultValue: any,
        public valueRepresentation: string) {
    }

}

